### ISSUE TITLE
False positive for TypeParameterUnusedInFormals

### ISSUE THREAD

**Issue by PhilippWendler** [google/error-prone#343]
I have the following method declaration in an interface:

```
  <R,
   R1 extends R,
   R2 extends R,
   X1 extends Exception,
   X2 extends Exception,
   V extends Visitor1<R1, X1> & Visitor2<R2, X2>
  >
  R accept_(V v) throws X1, X2;
```
(Explanation: I have two visitor interfaces, both parameterized in the return type and exception type of their methods, and this method should accept visitors that implement both interfaces.
The return type of this method is a super type of both visitors' return types.)

Error-prone 2.0.3 reports this warning:
```
[TypeParameterUnusedInFormals] Declaring a type parameter that is only used in the return type is a misuse of generics: operations on the type parameter are unchecked, it hides unsafe casts at invocations of the method, and it interacts badly with method overload resolution
    [javac]   R accept_(V v) throws X1, X2;
    [javac]     ^
    [javac]     (see http://errorprone.info/bugpattern/TypeParameterUnusedInFormals)
    [javac]   Did you mean '<R1 extends R,'?
```

First, I think my code is correct and does not contain a problem,
because while `R` is not used in the parameter types, it is bounded by `R1` and `R2`, both of which are in the parameter types (at least indirectly through V).

Second, the suggestion is not understandable (where should I put this `<R1 extends R,`? It is not valid where the arrow points to). The name of the type parameter could also be added to the warning text.

**Comment by cushon** [comment]
Thanks for the report, I think there are two bugs here. The check is only supposed to complain if the return type is not "reachable" from the formal parameter types, but as you mentioned they're connected through the bounds.

The fix is also broken. The diagnostics only show the first line of a fix. My guess is that the complete fix it's trying to suggest is:

```java
 <R1 extends R,
   R2 extends R,
   X1 extends Exception,
   X2 extends Exception,
   V extends Visitor1<R1, X1> & Visitor2<R2, X2>
  >
  R accept_(V v) throws X1, X2;
```

... but that still doesn't make any sense.
